app/services/katello/pulp3/smart_proxy_repository.rb

```diff
@@ -33,15 +33,45 @@ def current_repositories(environment_id = nil, content_view_id = nil)
         katello_repos.select { |repo| repo_ids.include? repo.pulp_id }
       end
 
+      def delete_dangling_distributions
+        distribution_deletion_tasks = []
+        distro_hrefs = []
+        orphan_repository_versions.each do |api, version_hrefs|
+          next if api.repository_type.id == :deb
+          if api.repository_type.pulp3_skip_publication
+            version_hrefs.collect do |href|
+              repo_href = href.split("/versions").first
+              distro_hrefs = RepositoryReference.where(repository_href: repo_href).collect do |repo_reference|
+                repos_id = RootRepository.where(id: repo_reference.root_repository_id).repositories.pluck(:id)
+                DistributionReference.where(repository_id: repos_id).collect(&:href)
+              end
+            end
+            distro_hrefs.flatten!
+            distribution_deletion_tasks << distro_hrefs.collect do |href|
+              api.delete_distribution(href)
+            end
+
+          else
+            distribution_deletion_tasks << version_hrefs.collect do |href|
+              api.publications_api.list({repository_version: href}).results.collect do |publication|
+                api.distributions_api.list({publication: publication.pulp_href}).results.collect do |distribution|
```

Comment:
I don't think any Pulp plugin API supports searching by publication. When I try it by hand, I'm seeing random things get returned that don't make sense. For example, I synced a Python repo twice and ran orphan cleanup. The distribution for the latest publication got deleted for some reason.

https://docs.pulpproject.org/pulp_python/restapi.html#tag/Distributions:-Pypi/operation/distributions_python_pypi_list

Even RPM doesn't seem to support listing distributions by publication.

I think we're going to have to either request the feature and wait or work around deleting these floating distributions.

Did you find that there are floating distributions often? I feel like this shouldn't be the case, since we delete distributions when repositories get deleted. I wonder if removing content view versions from an environment could cause that since the LCE distribution would no longer point to any repository.

Reply:
Ack..Removing this piece for now. Strangely the VCR re-recording ran without errors. There were some test repos left in pulp after the test run however.  Probably some tests that still don't have proper teardowns.